portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -20,92 +20,218 @@
-    int maxNumberOfComments;
-    try {
-        maxNumberOfComments = Integer.parseInt(request.getParameter("maxcomments"));
-    } catch (NumberFormatException e) {
-        maxNumberOfComments = MAX_COMMENTS_NUMBER;
-    }
-    if (maxNumberOfComments < 0 || maxNumberOfComments > MAX_COMMENTS_NUMBER) {
-        maxNumberOfComments = MAX_COMMENTS_NUMBER;
-    }
+        // get maxComments parameter
+        String maxNumberOfCommentsString = request.getParameter("maxcomments");
+        int maxNumberOfComments;
+        if (maxNumberOfCommentsString == null) {
+            maxNumberOfComments = MAX_COMMENTS_NUMBER;
+        } else {
+            try {
+                maxNumberOfComments = Integer.parseInt(maxNumberOfCommentsString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+        if (maxNumberOfComments < 0 || maxNumberOfComments > MAX_COMMENTS_NUMBER) {
+            response.sendRedirect("/");
+            return;
+        }
+
+        // get lastTimestamp parameter
+        String lastTimestampString = request.getParameter("timestamp");
+        long lastTimestamp;
+        if (lastTimestampString == null) {
+            lastTimestamp = System.currentTimeMillis() + 10;
+        } else {
+            try {
+                lastTimestamp = Long.parseLong(lastTimestampString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+
+        // get direction parameter
+        String direction = request.getParameter("direction");
+        if (direction == null) {
+            direction = "next";
+        }
+        if (!direction.equals("next") && !direction.equals("previous")) {
+            response.sendRedirect("/");
+            return;
+        }
+
+        // get commentsOnPage parameter (on current page)
+        String commentsOnPageString = request.getParameter("commentsonpage");
+        int commentsOnPage;
+        if (commentsOnPageString == null) {
+            commentsOnPage = maxNumberOfComments;
+        } else {
+            try {
+                commentsOnPage = Integer.parseInt(commentsOnPageString);
+            } catch (NumberFormatException e) {
+                response.sendRedirect("/");
+                return;
+            }
+        }
+
+        // query options
+        Filter timeFilter;
+        SortDirection sortDirection;
+        FetchOptions fetchOptions;
+
+        // set query options depending on whether the user want to see next or previous comments
+        if (direction.equals("next")) {
+            timeFilter =
+                    new FilterPredicate("timestamp", FilterOperator.LESS_THAN, lastTimestamp);
+            sortDirection = SortDirection.DESCENDING;
+            fetchOptions = FetchOptions.Builder.withLimit(maxNumberOfComments);
+        } else {
```

Comment:
Does 'direction' only take 'next' and 'previous' values? I think I saw 'stay' somewhere too. Is this condition branch valid for all other values, or only for the 'previous'?

Reply:
Here direction can be only "next" or "previous". It could change to "stay" a bit later.
I got rid of "stay" direction in commit to this PR. I don't know is the code more readable with or without "stay" direction.